docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,212 @@
+thread-safe map of `<sender:*Tx>`, i.e a mapping from `sender` to a single `*Tx`
+object, where each `*Tx` is the next valid and processable transaction from the
+given `sender`.
+
+On top of this mapping, we index all transactions by priority using a thread-safe
+priority queue, i.e. a [max heap](https://en.wikipedia.org/wiki/Min-max_heap).
+When a proposer is ready to select transactions for the next block proposal,
+transactions are selected from this priority index by highest priority order.
+When a transaction is selected and reaped, it is removed from this index and
+from the `<sender:*Tx>` mapping.
+
+We define `Tx` as the following data structure:
+
+```go
+type Tx struct {
+  // Tx represents the raw binary transaction data.
+  Tx []byte
+
+  // Priority defines the transaction's priority as specified by the application
+  // in the ResponseCheckTx response.
+  Priority int64
+
+  // Sender defines the transaction's sender as specified by the application in
+  // the ResponseCheckTx response.
+  Sender string
+
+  // Index defines the current index in the priority queue index. Note, if
+  // multiple Tx indexes are needed, this field will be removed and each Tx
+  // index will have its own wrapped Tx type.
+  Index int
+}
+```
+
+### Eviction
+
+Upon successfully executing `CheckTx` for a new `Tx` and the mempool is currently
+full, we must check if there exists a `Tx` of lower priority that can be evicted
+to make room for the new `Tx` with higher priority and with sufficient size
+capacity left.
+
+If such a `Tx` exists, we find it by obtaining a read lock and sorting the
+priority queue index. Once sorted, we find the first `Tx` with lower priority and
+size such that the new `Tx` would fit within the mempool's size limit. We then
+remove this `Tx` from the priority queue index as well as the `<sender:*Tx>`
+mapping.
+
+This will require additional `O(n)` space and `O(n*log(n))` runtime complexity.
+
+### Gossiping
+
+When gossiping new transactions to peers, we first obtain a read lock on the
+priority queue index. If we have not gossiped the top priority `Tx` for the
+given peer, we return this `Tx` as the next transaction to gossip. Otherwise,
+we sort the priority queue index and traverse in descending priority order until
+we find the first non-sent `Tx` to the peer.
+
+In order to facilitate this, we must also keep an ephemeral cache, mapping from
+`sender` to a list of unique `Tx` identifiers (e.g. the `Tx` hash). A `Tx`
+identifier is removed from this cache once it is either committed in a block or
+gossiped to a peer. Note, if the `Tx` is committed in a block, this requires that
+we iterates over the entire cache, i.e. each peer's list.
+
+## Consequences
+
+### Positive
+
+- Transactions are allowed to be prioritized by the application.
+- Transactions are allowed to be gossiped by priority.
+- Allows future replace-by-priority functionality.
+
+### Negative
```

Comment:
Another negative that I might consider adding is that before, when a transaction was submitted to a mempool and passed checkTx it was pretty much gauranteed to get in because nothing was evicted. Now with eviction it's possible that a transaction gets into the mempool and is evicted without the sender of the tx knowing that it was evicted.

Reply:
> Another negative that I might consider adding is that before, when a transaction was submitted to a mempool and passed checkTx it was pretty much gauranteed to get in because nothing was evicted.

This is not true AFAIK, the current mempool also has size limits so if a tx passes CheckTx but there isn't room it will not make it in.

However, with the new design it can make it in and then later be evicted. Yeah, I'll make a note of this.

Reply:
Wouldn't it check if there's room before sending it to the app 🤔 . In any case I was more coming from the perspective of using broadcastTxSync which would let you know if your tx had made it in the mempool and given a fifo system was almost gauranteed to be committed in a block. It's really a minor note but now if I do broadcastTxSync the gaurantee isn't so solid.